File: frontend/src/api/post.ts
import { graphql } from "../gql";

export const GET_POSTS = graphql(`
	query getPosts {
		posts {
			id
			content
			createdAt
			updatedAt
			user {
				id
				username
			}
			comments {
				id
			}
			likes {
				id
			}
		}
	}
`);

export const GET_POST = graphql(`
	query getPost($postId: ID!) {
		post(id: $postId) {
			id
			content
			createdAt
			updatedAt
			user {
				id
				username
			}
			comments {
				id
				content
				createdAt
				updatedAt
				user {
					id
					username
				}
			}
			likes {
				id
			}
		}
	}
`);

export const CREATE_POST = graphql(`
	mutation createPost($content: String!) {
		createPost(content: $content) {
			id
			content
		}
	}
`);

File: backend/internal/domain/post.go
package domain

import (
	"time"

	"github.com/uptrace/bun"
)

type Post struct {
	bun.BaseModel `bun:"table:posts,alias:p"`
	ID            int        `bun:"id,pk,autoincrement"`
	UserID        int        `bun:"user_id,notnull"`
	Content       string     `bun:"content,notnull"`
	CreatedAt     time.Time  `bun:"created_at,default:current_timestamp"`
	UpdatedAt     time.Time  `bun:"updated_at,default:current_timestamp"`
	User          *User      `bun:"rel:belongs-to,join:user_id=id"`
	Comments      []*Comment `bun:"rel:has-many,join:id=post_id"`
	Likes         []*Like    `bun:"rel:has-many,join:id=post_id"`
}

func WithComments(s *bun.SelectQuery) *bun.SelectQuery {
	return s.Relation("Comments")
}

func WithLikes(s *bun.SelectQuery) *bun.SelectQuery {
	return s.Relation("Likes")
}